NVIDIA/deepops
# Kubeflow 1.2

* Bump version to kubeflow v1.2 (manifests and kfctl)
* Fix GUI bug from v1.2 update. We can now set Tolerations and Affinities to Notebooks. This requires some configuration, but this will allow us to specify GPU architectures.

![image](https://user-images.githubusercontent.com/1010167/101970755-2cc94380-3be1-11eb-9b45-9cb71d396f0c.png)


* Add a new mechanism to the Kubeflow deployment script that allows kustomize overlays be saved in the workloads directory

Used the new mechanism to:
* Set the NGC example containers as the new defaults in Kubeflow
* Set the default Notebook configuration to use 1 GPU

![image](https://user-images.githubusercontent.com/1010167/101970697-d3611480-3be0-11eb-8d79-3c92e494055e.png)

* Add a default configuration example for notebook culling (default is set to off)
![image](https://user-images.githubusercontent.com/1010167/101970745-1ae7a080-3be1-11eb-98c6-7712cbd0e67d.png)


See https://github.com/NVIDIA/deepops/pull/760

Test plan:
Run the below basic verifications for the dex and istio kubeflow deployments.
1. Deploy Kubeflow
2. Launch all NGC containers, verify Jupyter
3. Set culling to true and idle_time to 1 (in the `.cache/...../deployment_patch.yaml`) & re-apply kubeflow (`cd ~/deepops/config/kubeflow-install && ../kfctl apply -f kfctl_k8s_istio.v1.2.0.yaml`)
4. Start a sleep container via pipelines, wait 4 minutes, and verify it shuts down.

> Note: It seems like the kubeflow culling only works for pipelines do to namespacing/permissions:
```
2020-12-12T02:53:29.714Z        INFO    controllers.Notebook    Updating container state:       {"notebook": "anonymous/bert", "namespace": "anonymous", "name": "bert"}
2020-12-12T02:53:29.733Z        INFO    culler  Warning: GET to http://bert.anonymous.svc.cluster.local/notebook/anonymous/bert/api/status: 403
2020-12-12T02:53:29.733Z        INFO    controllers.Notebook    Updating StatefulSet    {"notebook": "anonymous/newbiew12", "namespace": "anonymous", "name": "newbiew12"}
2020-12-12T02:53:29.741Z        INFO    controllers.Notebook    Updating container state:       {"notebook": "anonymous/newbiew12", "namespace": "anonymous", "name": "newbiew12"}
2020-12-12T02:53:29.752Z        INFO    culler  Warning: GET to http://newbiew12.anonymous.svc.cluster.local/notebook/anonymous/newbiew12/api/status: 403
2020-12-12T02:53:29.753Z        ERROR   controllers.Notebook    unable to fetch Notebook        {"notebook": "anonymous/bert-0.164fd734e05996ce", "error": "Notebook.kubeflow.org \"bert-0.164fd734e05996ce\" not found"}
```
